### PR TITLE
[feature] driver andorshrk: add support for controlling the iris and input flipper

### PR DESCRIPTION
**Pull request by pieleric**
The Kymera 328 supports a diaphragm iris, to control how much light comes in.
Add ability to control it if it's present.

The hardware unit is arbitrary (0->100). We convert it to something closer to reality.

Also introduce support for the input flipper (flip-in) axis, as the
Kymera 328 as such motorized input, and it's important to be able to
select the right input! Mostly just copy-paste from the flip-out axis.

### CODE REVIEW

**Comment by pieleric** [comment]
> what is an iris?

Essentially... another axis!
For more details, see: https://www.adorama.com/alc/iris-diaphragm/

**Comment by sneepstefan** [comment]
> > what is an iris?
> 
> Essentially... another axis! For more details, see: https://www.adorama.com/alc/iris-diaphragm/

Is it an idea to put this info and maybe a bit more elaborate in the description of the PR?

File: src/odemis/driver/andorshrk.py [hunk 1]
```diff
@@ -1777,6 +1872,33 @@ def _doSetSlitAbs(self, sid, width):
         self.SetAutoSlitWidth(sid, width)
         self._updatePosition()
 
+    def _doSetIrisRel(self, iris_id: int, shift: float):
+        """
+        Change the iris opening by a value
+        iris_id (int): iris ID
+        shift (float): change in opening size in m
```
Comment (tepals) [inline src/odemis/driver/andorshrk.py:1885]:
```
        :param iris_id: (int) iris ID
        :param shift: (float) change in opening size in m
```

File: src/odemis/driver/andorshrk.py [hunk 2]
```diff
@@ -1777,6 +1872,33 @@ def _doSetSlitAbs(self, sid, width):
...
+            raise ValueError("Position %f of axis '%s' not within range %f→%f" %
+                             (width, n, rng[0], rng[1]))
+
+        self._doSetIrisAbs(iris_id, width)
+
+    def _doSetIrisAbs(self, iris_id: int, width: float):
+        """
+        Change the iris opening to a value
+        iris_id (int): iris ID
+        width (float): new position in m
```
Comment (tepals) [inline src/odemis/driver/andorshrk.py:1902]:
```
        :param iris_id: (int) iris ID
        :param width: (float) new position in m
```

File: src/odemis/driver/andorshrk.py [hunk 3]
```diff
@@ -317,6 +330,7 @@ def __init__(self, name, role, device, camera=None, accessory=None,
         slitleds_settle_time (0 <= float): duration wait before (potentially)
           turning on the slit leds. Useful to delay the move after the slitleds
           interlock is set.
+        iris (None, or dict int -> str): for each iris (0=direct, 1=side) -> axis name.
```
Comment (tepals) [inline src/odemis/driver/andorshrk.py]:
```
        :param iris: (None, or dict int -> str) for each iris (0=direct, 1=side) -> axis name.

```

File: src/odemis/driver/andorshrk.py [hunk 4]
```diff
@@ -1972,11 +2094,14 @@ def __init__(self, ccd=None):
...
                       }
         # flippers: int (id) -> int (port number, 0 or 1)
-        self._flippers = {2: 0}
+        self._flippers = {INPUT_FLIPPER: DIRECT_PORT, OUTPUT_FLIPPER: DIRECT_PORT}
 
         # accessory: 2 lines -> int (0 or 1)
         self._accessory = [0, 0]
 
+        # iris: 2 iris (DIRECT_PORT, SIDE_PORT) -> int (0->100)
+        self._iris = [10, 50]
```
Comment (tepals) [inline src/odemis/driver/andorshrk.py:2109]: shouldn't this be a dict?

**Review by tepals — CHANGES_REQUESTED** [review]

File: src/odemis/driver/andorshrk.py [hunk 5]
```diff
@@ -295,47 +307,55 @@ class Shamrock(model.Actuator):
...
+        much the image will change after a move.
+        :param drives_shutter: flip-out angles for which the shutter should be set
+        to BNC (external) mode. Otherwise, the shutter is left opened.
+        :param dependencies (None or dict str -> HwComponent): if the key starts with
+        "led_prot", it will set the .protection to True any time that the
+        slit leds could be turned on.
         :param check_move: Name of the axis -> bool: If True (default), after move, raise an error if
         that position is not the expected one. If False, if the position is not the expected one,
         just log a warning. Note: for now, only some axes are actually checked (flipper and band)
+        :param inverted: Must be None (or not passed), as it is not allowed to invert the axes.
```
Comment (sneepstefan) [inline src/odemis/driver/andorshrk.py:358]: param inverted is not an explicit argument, is it required to be included here? You could also move it ( as a non-argument description) to line 366.

**Review by sneepstefan — CHANGES_REQUESTED** [review]

File: src/odemis/driver/andorshrk.py [hunk 6]
```diff
@@ -276,6 +276,18 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+IRIS_NAMES = {
+    DIRECT_PORT: "iris-direct",
+    SIDE_PORT: "iris-side",
+}
+
+# Conversion factor from the arbitrary values to the actual opening in m, roughly.
+# Experimentally, the full opening of the iris is ~20mm. The minimum opening is > 0 (there is still light).
+IRIS_OPENING_TO_M = 200e-6  # m, approximate
+IRIS_POS_MIN = 0
+IRIS_POS_MAX = 100
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/andorshrk.py:289]: Yes, indeed, it's not really arbitrary. It's just a percentage between the minimum opening and maximum opening. 0 is *not* closed though. I'll update the comment.

File: src/odemis/driver/andorshrk.py [hunk 7]
```diff
@@ -1423,10 +1507,15 @@ def _updatePosition(self, must_notify=False):
             if name in self.axes:
                 pos[name] = self.GetAutoSlitWidth(i)
 
-        if "flip-out" in self.axes:
-            val = self.GetFlipperMirror(OUTPUT_FLIPPER)
-            userv = [k for k, v in FLIPPER_TO_PORT.items() if v == val][0]
-            pos["flip-out"] = userv
+        for i, name in self._iris_names.items():
+            if name in self.axes:
+                pos[name] = (self.GetIris(i) + 1) * IRIS_OPENING_TO_M
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/andorshrk.py:1518]: Indeed. I've now made 2 conversion functions iris_hw_to_user() and iris_user_to_hw()

File: src/odemis/driver/andorshrk.py (hunk 3 above)
Comment (pieleric) [reply src/odemis/driver/andorshrk.py]: That's fixed in the second commit.

File: src/odemis/driver/andorshrk.py (hunk 5 above)
Comment (pieleric) [reply src/odemis/driver/andorshrk.py:358]: It's not explicit in the signature, but it's part of the Actuator signature, and it's handled explicitly in the code. So I think it's worthy to be placed in the docstring.

**Review by sneepstefan — APPROVED** [review]

**Review by tepals — APPROVED** [review]

**Pull request merged**